### DISCUSSION TITLE
Remote Tile Server

### DISCUSSION THREAD

**Discussion by banesullivan** [banesullivan/localtileserver#32] [Show and tell]
I haven't documented this in the README (yet), but in #30, I recently added support for using a remotely deployed instance of `localtileserver` with URL files.

I have a template repository in case someone wants to spin up their own: https://github.com/banesullivan/remotetileserver

Note that the backing flask app here does not do any sort of authentication, so once `localtileserver` is deployed, anyone can access it.

But once you have a remote instance running, or you want to try out my demo server, you can simply do:

```py
from localtileserver import get_leaflet_tile_layer
from localtileserver import RemoteTileClient
from ipyleaflet import Map

# Connect to a remote instance of localtileserver
tile_client = RemoteTileClient(
    'https://opendata.digitalglobe.com/events/california-fire-2020/pre-event/2018-02-16/pine-gulch-fire20/1030010076004E00.tif',
    #'https://localtileserver-demo.herokuapp.com/',
)

# Create folium tile layer from that server
t = get_leaflet_tile_layer(tile_client)

m = Map(center=tile_client.center())
m.add_layer(t)
m
```

Similarly, pass any URL file to `RemoteTileClient` and it will use the server to load and deliver tiles. Again, COGs are best when doing everything remotely like this -- and actually quite efficient

This provides a workaround for #29 (cc @giswqs)

**Comment by banesullivan** [comment]
Here are some sample URL files to test this out with:

- `https://opendata.digitalglobe.com/events/california-fire-2020/pre-event/2018-02-16/pine-gulch-fire20/1030010076004E00.tif`
- `s3://sentinel-cogs/sentinel-s2-l2a-cogs/2020/S2A_31QHU_20200714_0_L2A/B01.tif`

**Comment by giswqs** [comment]
This is superb! Looking forward to trying it out. Thanks a lot for making it happen

**Comment by banesullivan** [comment]
I've posted a MyBinder demo using all of this in this repo: https://github.com/banesullivan/localtileserver-demo

Which is launched here: https://mybinder.org/v2/gh/banesullivan/localtileserver-demo/HEAD

**Reply by CChisuwo** [reply]
Hello! I think this is an amazing project, I'm definitely using it in several of my projects. I was however wondering if it is possible to launch a tile server, generate a url and be able to use that in other applications like qgis or other tile maps

**Reply by banesullivan** [reply]
That is definitely possible. I'd recommend launching `localtileserver` locally from the command line like:

```bash
$ localtileserver --browser False path/to/raster.tif
 * Serving Flask app 'localtileserver.tileserver.application' (lazy loading)
 * Environment: production
   WARNING: This is a development server. Do not use it in a production deployment.
   Use a production WSGI server instead.
 * Debug mode: off
 * Running on http://127.0.0.1:53931 (Press CTRL+C to quit)
```

Then use the output URL, in my example it is `http://127.0.0.1:53931`, to go to the API docs at `http://127.0.0.1:53931/swagger`. There you will find the available REST endpoints. Specifically, you will want to use the `​/tiles​/{z}​/{x}​/{y}.png` endpoint when creating a new TMS tile layer in your application of choice.

In my example, the URL would be: `http://127.0.0.1:53931​/tiles​/{z}​/{x}​/{y}.png`

In an application like QGIS, you will create a new TMS tile layer passing that URL template. I do not use QGIS, but this guide seems to be appropriate: https://felix.rohrba.ch/en/2017/easily-add-tilemap-layers-qgis/

I hope this helps.

FYI: by default, `localtileserver` will launch on a random available port but you can control the port so that you do not have to recreate the TMS source in QGIS everytime the server relaunches with the `--port` flag. For example:

```bash
localtileserver --browser False --port 8088 path/to/raster.tif
```

**Reply by nikihikko** [reply]
Hi, I did the actions you described above, but I failed, I started the server and I can go to it and see the map, but when I add it to qgis or leaflet using the link, nothing happens.

I installed localtileserver via miniconda command: `conda install -c conda-forge localtileserver ipyleaflet`
ran server command: `localtileserver --browser False --port 5377 ./cog_TC_NG_SFBay_US_Geo_small.tif`
as a result, at `http://127.0.0.1:5377/` I see a map!
then I add the map to qgis like this:
https://drive.google.com/file/d/1EjwmBoR3uKgw1hH0ArI-sUeBWOYSLklH/view?usp=share_link
what I see in when connecting this map to qgis:
https://drive.google.com/file/d/1L64u2Q8EQkeNxcnNMH5K_8g4xxpkwvCf/view?usp=share_link
what I’d like to see:
https://drive.google.com/file/d/1HJ50PhLQBz8L4iXjdta-ZuOKNilMPLks/view?usp=share_link
console log: https://pastebin.com/CR7qeBb1

Please tell me what I’m doing wrong and how to fix it?

**Reply by banesullivan** [reply]
@nikihikko, sorry for my delay.. would you please open a new discussion topic for this?

**Comment by Umarsharief-07** [comment]
hi can you tell how can i use this titleserver i have deployed my application in conatiner but the problem arriess its lunching localhost 127.0.0.1:<random_port> so because of this the formation of tiff is not possible in web . fyi  i have put envs like this import os
os.environ["LOCALTILESERVER_PORT"] = "8090"
os.environ["LOCALTILESERVER_HOST"] = "0.0.0.0" # Binds publicly within the container
os.environ["LOCALTILESERVER_CLIENT_PORT"] = "8090"
os.environ["LOCALTILESERVER_CLIENT_HOST"] = "13.200.222.85" # Public IP
os.environ["LOCALTILESERVER_CLIENT_PREFIX"] = ""

but in container its like 127.0.0.1:<random_ip>
